test/e2e/operator/nmstate_install_test.go

```diff
@@ -0,0 +1,135 @@
+	err := framework.Global.Client.Get(context.TODO(), types.NamespacedName{Namespace: webhook.ObjectMeta.Namespace, Name: webhook.ObjectMeta.Name}, &webhook)
+	return webhook, err
+}
+
+func getNMStateHandlerList(handler v1.DaemonSet) (v1.DaemonSetList, error) {
+	ds := v1.DaemonSetList{}
+	err := framework.Global.Client.List(context.TODO(), &ds, &client.ListOptions{Namespace: handler.ObjectMeta.Namespace})
+	return ds, err
+}
+
+var _ = Describe("NMState operator", func() {
+	var (
+		nmstate nmstatev1beta1.NMState
+		handler v1.DaemonSet
+		webhook v1.Deployment
+	)
+	BeforeEach(func() {
+		nmstate = nmstatev1beta1.NMState{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "nmstate",
+				Namespace: "nmstate",
+			},
+		}
+		handler = v1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "nmstate-handler",
+				Namespace: "nmstate",
+			},
+		}
+		webhook = v1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "nmstate-webhook",
+				Namespace: "nmstate",
+			},
+		}
+	})
+	Context("when installed for the first time", func() {
+		BeforeEach(func() {
+			installNMState(nmstate)
+		})
+		It("should deploy daemonset and webhook deployment", func() {
+			Eventually(func() bool {
+				_, err := getNMStateHandler(handler)
+				return apierrors.IsNotFound(err)
+			}, 60*time.Second, 1*time.Second).ShouldNot(BeTrue(), fmt.Sprint("Daemon set for NMState should be found"))
+			Eventually(func() bool {
+				_, err := getNMStateWebhook(webhook)
+				return apierrors.IsNotFound(err)
+			}, 60*time.Second, 1*time.Second).ShouldNot(BeTrue(), fmt.Sprint("Deployment for NMState webhook should be found"))
+		})
+	})
+	Context("when NMState is installed", func() {
+		BeforeEach(func() {
+			ds, err := getNMStateHandlerList(handler)
+			Expect(err).ToNot(HaveOccurred(), "List daemon sets in namespace nmstate succeeds")
+			Expect(ds.Items).To(HaveLen(1), "and has only one item")
+		})
+		Context("and the CR is created with a wrong name", func() {
+			BeforeEach(func() {
+				nmstate.SetName("nmstate-wrong")
+				err := framework.Global.Client.Create(context.TODO(), &nmstate, &framework.CleanupOptions{})
+				Expect(err).ToNot(HaveOccurred(), "NMState CR with incorrect name is created without error")
+			})
+			AfterEach(func() {
+				nmstate.SetName("nmstate-wrong")
+				err := framework.Global.Client.Delete(context.TODO(), &nmstate, &client.DeleteOptions{})
+				Expect(err).ToNot(HaveOccurred(), "NMState CR with incorrect name is removed without error")
+			})
+			It("should ignore it", func() {
+				ds, err := getNMStateHandlerList(handler)
+				Expect(err).ToNot(HaveOccurred(), "Daemon set list is retreieved without error")
+				Expect(ds.Items).To(HaveLen(1), "and still only has one item")
+			})
+		})
+	})
+	Context("when NMState is uninstalled", func() {
+		BeforeEach(func() {
+			uninstallNMState(nmstate)
+		})
+		It("should uninstall handler and webhook", func() {
+			Eventually(func() bool {
+				_, err := getNMStateHandler(handler)
```

Comment:
Do we delete only handler DS and webhook deployment when operator CR is removed?
Maybe it worth to add a test that checks: Create nsmatte, delete nmstate, and create nmstate once again, to verify that we don't fail due to leftovers?  but it could be in a follow up patch

Reply:
That would definitely be in scope for a follow-up. There are some other resources that should be deleted when the handler is removed.